File: backend-go-fiber/src/controllers/article.go
package controllers

import (
	"context"
	"time"

	"backend-go-fiber/src/database"
	"backend-go-fiber/src/models"

	"github.com/gofiber/fiber/v2"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo/options"
)

type ArticleController struct{}

// GetAllArticles returns list of articles with optional filters
func (ac *ArticleController) GetAllArticles(c *fiber.Ctx) error {
	lang := c.Query("lang", "id")
	status := c.Query("status")
	category := c.Query("category")

	ctx := context.Background()
	articleCollection := database.GetCollection("articles")

	// Build filter
	filter := bson.M{}
	
	// Only filter by status if explicitly provided and not empty string
	if status != "" {
		filter["status"] = status
	}
	
	if category != "" {
		filter["category"] = category
	}

	// Find articles
	opts := options.Find().SetSort(bson.D{{Key: "publishedAt", Value: -1}})
	cursor, err := articleCollection.Find(ctx, filter, opts)
	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"success": false,
			"message": "Error fetching articles",
			"error":   err.Error(),
		})
	}
	defer cursor.Close(ctx)

	var articles []models.Article
	if err := cursor.All(ctx, &articles); err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"success": false,
			"message": "Error parsing articles",
			"error":   err.Error(),
		})
	}

	// Transform articles to previews with requested language
	var previews []models.ArticlePreview
	includeStatus := status == "" // Include status if fetching all articles
	
	for _, article := range articles {
		if preview := article.GetPreview(lang, includeStatus); preview != nil {
			previews = append(previews, *preview)
		}
	}

	return c.Status(fiber.StatusOK).JSON(fiber.Map{
		"success": true,
		"count":   len(previews),
		"data":    previews,
	})
}

// GetArticleBySlug returns article detail by slug
func (ac *ArticleController) GetArticleBySlug(c *fiber.Ctx) error {
	slug := c.Params("slug")
	lang := c.Query("lang", "id")

	ctx := context.Background()
	articleCollection := database.GetCollection("articles")

	// Find article by slug and language
	filter := bson.M{
		"translations": bson.M{
			"$elemMatch": bson.M{
				"slug": slug,
				"lang": lang,
			},
		},
	}

	var article models.Article
	err := articleCollection.FindOne(ctx, filter).Decode(&article)
	if err != nil {
		return c.Status(fiber.StatusNotFound).JSON(fiber.Map{
			"success": false,
			"message": "Article not found",
		})
	}

	// Get article detail for requested language
	detail := article.GetByLanguage(lang)
	if detail == nil {
		return c.Status(fiber.StatusNotFound).JSON(fiber.Map{
			"success": false,
			"message": "Article not found for this language",
		})
	}

	return c.Status(fiber.StatusOK).JSON(fiber.Map{
		"success": true,
		"data":    detail,
	})
}

// GetArticleByID returns article by ID with all languages
func (ac *ArticleController) GetArticleByID(c *fiber.Ctx) error {
	id := c.Params("id")

	objID, err := primitive.ObjectIDFromHex(id)
	if err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"success": false,
			"message": "Invalid article ID",
		})
	}

	ctx := context.Background()
	articleCollection := database.GetCollection("articles")

	var article models.Article
	err = articleCollection.FindOne(ctx, bson.M{"_id": objID}).Decode(&article)
	if err != nil {
		return c.Status(fiber.StatusNotFound).JSON(fiber.Map{
			"success": false,
			"message": "Article not found",
		})
	}

	return c.Status(fiber.StatusOK).JSON(fiber.Map{
		"success": true,
		"data":    article,
	})
}

// CreateArticle creates a new article
func (ac *ArticleController) CreateArticle(c *fiber.Ctx) error {
	var article models.Article
	if err := c.BodyParser(&article); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"success": false,
			"message": "Invalid request body",
			"error":   err.Error(),
		})
	}

	// Set timestamps
	article.ID = primitive.NewObjectID()
	article.CreatedAt = time.Now()
	article.UpdatedAt = time.Now()

	if article.PublishedAt.IsZero() {
		article.PublishedAt = time.Now()
	}

	ctx := context.Background()
	articleCollection := database.GetCollection("articles")

	_, err := articleCollection.InsertOne(ctx, article)
	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"success": false,
			"message": "Error creating article",
			"error":   err.Error(),
		})
	}

	return c.Status(fiber.StatusCreated).JSON(fiber.Map{
		"success": true,
		"message": "Article created successfully",
		"data":    article,
	})
}

// UpdateArticle updates an existing article
func (ac *ArticleController) UpdateArticle(c *fiber.Ctx) error {
	id := c.Params("id")

	objID, err := primitive.ObjectIDFromHex(id)
	if err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"success": false,
			"message": "Invalid article ID",
		})
	}

	var updateData models.Article
	if err := c.BodyParser(&updateData); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"success": false,
			"message": "Invalid request body",
			"error":   err.Error(),
		})
	}

	// Update timestamp
	updateData.UpdatedAt = time.Now()

	ctx := context.Background()
	articleCollection := database.GetCollection("articles")

	// Update article
	update := bson.M{
		"$set": bson.M{
			"translations": updateData.Translations,
			"category":     updateData.Category,
			"author":       updateData.Author,
			"coverImage":   updateData.CoverImage,
			"status":       updateData.Status,
			"updatedAt":    updateData.UpdatedAt,
		},
	}

	opts := options.FindOneAndUpdate().SetReturnDocument(options.After)
	result := articleCollection.FindOneAndUpdate(ctx, bson.M{"_id": objID}, update, opts)

	var article models.Article
	if err := result.Decode(&article); err != nil {
		return c.Status(fiber.StatusNotFound).JSON(fiber.Map{
			"success": false,
			"message": "Article not found",
		})
	}

	return c.Status(fiber.StatusOK).JSON(fiber.Map{
		"success": true,
		"message": "Article updated successfully",
		"data":    article,
	})
}

// DeleteArticle deletes an article
func (ac *ArticleController) DeleteArticle(c *fiber.Ctx) error {
	id := c.Params("id")

	objID, err := primitive.ObjectIDFromHex(id)
	if err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"success": false,
			"message": "Invalid article ID",
		})
	}

	ctx := context.Background()
	articleCollection := database.GetCollection("articles")

	result, err := articleCollection.DeleteOne(ctx, bson.M{"_id": objID})
	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"success": false,
			"message": "Error deleting article",
			"error":   err.Error(),
		})
	}

	if result.DeletedCount == 0 {
		return c.Status(fiber.StatusNotFound).JSON(fiber.Map{
			"success": false,
			"message": "Article not found",
		})
	}

	return c.Status(fiber.StatusOK).JSON(fiber.Map{
		"success": true,
		"message": "Article deleted successfully",
	})
}

// GetArticlesByCategory returns articles filtered by category
func (ac *ArticleController) GetArticlesByCategory(c *fiber.Ctx) error {
	category := c.Params("category")
	lang := c.Query("lang", "id")

	ctx := context.Background()
	articleCollection := database.GetCollection("articles")

	// Find published articles in category
	filter := bson.M{
		"category": category,
		"status":   "published",
	}

	opts := options.Find().SetSort(bson.D{{Key: "publishedAt", Value: -1}})
	cursor, err := articleCollection.Find(ctx, filter, opts)
	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"success": false,
			"message": "Error fetching articles by category",
			"error":   err.Error(),
		})
	}
	defer cursor.Close(ctx)

	var articles []models.Article
	if err := cursor.All(ctx, &articles); err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"success": false,
			"message": "Error parsing articles",
			"error":   err.Error(),
		})
	}

	// Transform to previews
	var previews []models.ArticlePreview
	for _, article := range articles {
		if preview := article.GetPreview(lang, false); preview != nil {
			previews = append(previews, *preview)
		}
	}

	return c.Status(fiber.StatusOK).JSON(fiber.Map{
		"success": true,
		"count":   len(previews),
		"data":    previews,
	})
}

File: backend-rust/src/controllers/article.rs
use actix_web::{web, HttpResponse};
use mongodb::{
    bson::{doc, oid::ObjectId},
    Database,
};
use serde::{Deserialize, Serialize};

use crate::{
    error::{AppError, AppResult},
    models::{Article, ArticleDetail, ArticlePreview},
};

#[derive(Deserialize)]
pub struct ArticleQuery {
    #[serde(default = "default_lang")]
    pub lang: String,
    
    #[serde(default = "default_status")]
    pub status: String,
    
    pub category: Option<String>,
}

fn default_lang() -> String {
    "id".to_string()
}

fn default_status() -> String {
    "published".to_string()
}

#[derive(Serialize)]
struct ArticlesResponse {
    success: bool,
    count: usize,
    data: Vec<ArticlePreview>,
}

#[derive(Serialize)]
struct ArticleResponse {
    success: bool,
    data: ArticleDetail,
}

#[derive(Serialize)]
struct ArticleFullResponse {
    success: bool,
    data: Article,
}

#[derive(Serialize)]
struct MessageResponse {
    success: bool,
    message: String,
}

#[derive(Serialize)]
struct CreateArticleResponse {
    success: bool,
    message: String,
    data: Article,
}

#[derive(Serialize)]
struct UpdateArticleResponse {
    success: bool,
    message: String,
    data: Article,
}

/// GET /api/articles
/// Get all articles with optional filters
pub async fn get_all_articles(
    db: web::Data<Database>,
    query: web::Query<ArticleQuery>,
) -> AppResult<HttpResponse> {
    let articles_collection = db.collection::<Article>("articles");
    
    // Build filter
    let mut filter = doc! {};
    if !query.status.is_empty() {
        filter.insert("status", &query.status);
    }
    if let Some(ref category) = query.category {
        filter.insert("category", category);
    }
    
    // Find articles
    let mut cursor = articles_collection
        .find(
            filter,
        )
        .sort(doc! { "publishedAt": -1 })
        .await?;
    
    let mut articles = Vec::new();
    while cursor.advance().await? {
        let article = cursor.deserialize_current()?;
        if let Some(preview) = article.to_preview(&query.lang) {
            articles.push(preview);
        }
    }
    
    let response = ArticlesResponse {
        success: true,
        count: articles.len(),
        data: articles,
    };
    
    Ok(HttpResponse::Ok().json(response))
}

/// GET /api/articles/slug/{slug}
/// Get article by slug
pub async fn get_article_by_slug(
    db: web::Data<Database>,
    slug: web::Path<String>,
    query: web::Query<ArticleQuery>,
) -> AppResult<HttpResponse> {
    let articles_collection = db.collection::<Article>("articles");
    
    // Find article with matching slug and language
    let article = articles_collection
        .find_one(
            doc! {
                "translations.slug": slug.as_str(),
                "translations.lang": &query.lang
            },
        )
        .await?
        .ok_or_else(|| AppError::NotFound("Article not found".to_string()))?;
    
    let article_data = article
        .get_by_language(&query.lang)
        .ok_or_else(|| AppError::NotFound("Translation not found".to_string()))?;
    
    let response = ArticleResponse {
        success: true,
        data: article_data,
    };
    
    Ok(HttpResponse::Ok().json(response))
}

/// GET /api/articles/{id}
/// Get article by ID (with all languages)
pub async fn get_article_by_id(
    db: web::Data<Database>,
    id: web::Path<String>,
) -> AppResult<HttpResponse> {
    let articles_collection = db.collection::<Article>("articles");
    
    // Parse ObjectId
    let article_id = ObjectId::parse_str(id.as_str())
        .map_err(|_| AppError::BadRequest("Invalid article ID".to_string()))?;
    
    // Find article
    let saved_article = articles_collection
        .find_one(doc! { "_id": article_id })
        .await?
        .ok_or_else(|| AppError::NotFound("Article not found".to_string()))?;
    
    let response = ArticleFullResponse {
        success: true,
        data: saved_article,
    };
    
    Ok(HttpResponse::Ok().json(response))
}

/// POST /api/articles
/// Create new article
pub async fn create_article(
    db: web::Data<Database>,
    req: web::Json<Article>,
) -> AppResult<HttpResponse> {
    let articles_collection = db.collection::<Article>("articles");
    
    // Validate that at least one translation exists
    if req.translations.is_empty() {
        return Err(AppError::ValidationError(
            "At least one translation is required".to_string()
        ));
    }
    
    // Insert article
    let insert_result = articles_collection.insert_one(req.into_inner()).await?;
    
    // Fetch the created article
    let article_id = insert_result.inserted_id.as_object_id()
        .ok_or_else(|| AppError::InternalError("Failed to get article ID".to_string()))?;
    
    let article = articles_collection
        .find_one(doc! { "_id": article_id })
        .await?
        .ok_or_else(|| AppError::InternalError("Failed to fetch created article".to_string()))?;
    
    let response = CreateArticleResponse {
        success: true,
        message: "Article created successfully".to_string(),
        data: article,
    };
    
    Ok(HttpResponse::Created().json(response))
}

/// PUT /api/articles/{id}
/// Update article
pub async fn update_article(
    db: web::Data<Database>,
    id: web::Path<String>,
    req: web::Json<Article>,
) -> AppResult<HttpResponse> {
    let articles_collection = db.collection::<Article>("articles");
    
    // Parse ObjectId
    let article_id = ObjectId::parse_str(id.as_str())
        .map_err(|_| AppError::BadRequest("Invalid article ID".to_string()))?;
    
    // Validate that at least one translation exists
    if req.translations.is_empty() {
        return Err(AppError::ValidationError(
            "At least one translation is required".to_string()
        ));
    }
    
    // Update article
    let update_doc = mongodb::bson::to_document(&req.into_inner())
        .map_err(|e| AppError::InternalError(format!("Failed to serialize: {}", e)))?;
    
    let updated_article = articles_collection
        .find_one_and_update(
            doc! { "_id": article_id },
            doc! { "$set": update_doc },
        )
        .return_document(mongodb::options::ReturnDocument::After)
        .await?
        .ok_or_else(|| AppError::NotFound("Article not found".to_string()))?;
    
    let response = UpdateArticleResponse {
        success: true,
        message: "Article updated successfully".to_string(),
        data: updated_article,
    };
    
    Ok(HttpResponse::Ok().json(response))
}

/// DELETE /api/articles/{id}
/// Delete article
pub async fn delete_article(
    db: web::Data<Database>,
    id: web::Path<String>,
) -> AppResult<HttpResponse> {
    let articles_collection = db.collection::<Article>("articles");
    
    // Parse ObjectId
    let article_id = ObjectId::parse_str(id.as_str())
        .map_err(|_| AppError::BadRequest("Invalid article ID".to_string()))?;
    
    // Delete article
    let result = articles_collection
        .delete_one(doc! { "_id": article_id })
        .await?;
    
    if result.deleted_count == 0 {
        return Err(AppError::NotFound("Article not found".to_string()));
    }
    
    let response = MessageResponse {
        success: true,
        message: "Article deleted successfully".to_string(),
    };
    
    Ok(HttpResponse::Ok().json(response))
}

/// GET /api/articles/category/{category}
/// Get articles by category
pub async fn get_articles_by_category(
    db: web::Data<Database>,
    category: web::Path<String>,
    query: web::Query<ArticleQuery>,
) -> AppResult<HttpResponse> {
    let articles_collection = db.collection::<Article>("articles");
    
    // Find articles by category
    let mut cursor = articles_collection
        .find(
            doc! {
                "category": category.as_str(),
                "status": "published"
            },
        )
        .sort(doc! { "publishedAt": -1 })
        .await?;
    
    let mut articles = Vec::new();
    while cursor.advance().await? {
        let article = cursor.deserialize_current()?;
        if let Some(preview) = article.to_preview(&query.lang) {
            articles.push(preview);
        }
    }
    
    let response = ArticlesResponse {
        success: true,
        count: articles.len(),
        data: articles,
    };
    
    Ok(HttpResponse::Ok().json(response))
}
